### ISSUE TITLE
Feature Request: Combined reducers

### ISSUE THREAD

**Issue by dhollman** [kokkos/kokkos#2013]
It should be possible to trivially combine reducers without having to write a custom reducer.  For instance, something like

```c++
Kokkos::parallel_reduce(n,
  KOKKOS_LAMBDA (int i, Kokkos::pair<double, int>& output) { /* ... */ },
  Kokkos::make_combined_reducer(Kokkos::Max<double>{x}, Kokkos::Sum<int>{y})
)
```

**Comment by crtrott** [comment]
Interesting thought, and shouldn't be too hard to do.

**Comment by mhoemmen** [comment]
Which argument goes first?  Is this like Ranges pipes?

**Comment by dhollman** [comment]
@mhoemmen I'm not sure I understand the question.  The reduction operation is just the operation of each reducer on the corresponding entries in the pair/tuple. What do you mean by "goes first"?

**Comment by mhoemmen** [comment]
@dhollman Sorry, I was confused about what `make_combined_reducer(Max, Sum)` does.

**Comment by crtrott** [comment]
Should this be a tuple instead, to allow to combine more than two?

**Comment by dhollman** [comment]
Yes, but honestly `pair` should just be a fancy alias for a tuple with two elements.  Also, we'd need to write a `Kokkos::tuple` (or  a `stdoe::tuple`?) for this to work.

**Comment by tomdeakin** [comment]
This would be super handy for me! I'd like to reduce more than 2 variables, but for me the reduction type is all the same e.g sum up X different things. The behaviour would be akin to an OpenMP `reduction(+:x,y,z)` clause.

**Comment by dhollman** [comment]
It was suggested at the Kokkos bootcamp in Oakland that an even more straightforward way of doing this would be to take a variadic set of reducers require the user to provide a parameter for each reducer.  This avoids the `tuple` problem altogether, though it's a bit more intrusive for us to implement.

**Comment by dholladay00** [comment]
@dhollman could you clarify a bit?
Is that what you mean? 
```c++
Kokkos::parallel_reduce(n,
  KOKKOS_LAMBDA (int i, double& mx, int& sy) { /* ... */ },
  Kokkos::make_combined_reducer(Kokkos::Max<double>{x}, Kokkos::Sum<int>{y})
)
```

**Comment by crtrott** [comment]
I think if we do the variadic thing, we wouldn't even do a combined reducer. I.e. you would simply do this:
```c++
Kokkos::parallel_reduce(n,
  KOKKOS_LAMBDA (int i, double& mx, int& sy) { /* ... */ },
  Kokkos::Max<double>{x}, Kokkos::Sum<int>{y}
);
```

**Comment by crtrott** [comment]
The tuple variant is simpler to implement for us, the variadic variant is less typing for folks. One thing we could consider is using the tuple variant as the implementation mechanism and having the variadic variant sit on top of it.

**Comment by dholladay00** [comment]
Wouldn't that require a `Kokkos::tuple` implementation or would you borrow thrust's?

**Comment by crtrott** [comment]
Yeah we would probably need a `Kokkos::tuple` implementation

**Comment by dhollman** [comment]
The CUDA 9.2 documentation *claims* that `std::tuple` just works out of the box.

**Comment by dhollman** [comment]
(They've special-cased it in the compiler to not need `__device__` markings)

**Comment by dhollman** [comment]
<https://docs.nvidia.com/cuda/archive/9.2/cuda-toolkit-release-notes/index.html#cuda-compiler-new-features>

**Comment by dhollman** [comment]
But regardless, I don't want our users to have to deal with `tuple` directly.  Packing and unpacking `tuple`s is a real pain in pre-C++17 code.  I don't think people should have to deal with it, especially when it's not that much harder to add the variadic version while we clean up the rest of the `parallel_reduce` stuff.

**Comment by mhoemmen** [comment]
It's fun _once_ to reimplement Lisp in C++, but less fun N >> 1 times ;-)

**Comment by pgrete** [comment]
:+1:  on this and #2045 
We'd also be interested in seeing a clean version as we now need this feature in our code.

**Comment by dalg24** [comment]
Addressed in #3052 and available since Kokkos 3.2
